src/FileBrowser.js

```diff
@@ -0,0 +1,230 @@
+'use strict';
+/* API for all file-browser commands */
+
+const
+  assert    = require('assert'),
+  debug     = require('debug')('browser'),
+  Buffer    = require('buffer').Buffer,
+  msp       = require('msgpack-lite'),
+  through2  = require('through2'),
+  Promise   = require('bluebird'),
+  lock      = require('lock')(),
+  _         = require('lodash'),
+  slugid    = require('slugid'),
+  fs        = require('fs');
+
+const
+  Command   = require('./Command.js');
+
+const CHUNKSIZE = 4072;
+
+const lck = key => new Promise(resolve => lock(key, resolve));
+
+class FileBrowser {
+  
+  constructor(inStream, outStream) {
+
+    assert(inStream);
+    assert(outStream);
+
+    this.stdin = through2.obj(function (chunk, enc, cb) {
+      this.push(msp.encode(chunk));
+      cb();
+    });
+
+    // Table that maps from id to resolver
+    this._cb = {}
+
+    this.stdout = msp.createDecodeStream();
+
+    this.stdin.pipe(inStream);
+    outStream.pipe(this.stdout);
+
+    this.stdout.on('data', (data) => {
+      if(!data.id) return;
+      debug('Received: ', data);
+      if (!this._cb[data.id]) {
+        debug("message for unknown id: ", data.id);
+        return;
+      }
+
+      this._cb[data.id](data);
+    });
+
+    this.stdout.on('error', debug);
+    this.stdin.on('error', debug);
+
+    let self = this;
+
+    ["mv", "cp"].forEach(c => {
+
+      self[c] = async (src, dest) => {
+        let cmd = Command[c](src, dest);
+        let result = await self._writeAndResolve(cmd);
+        delete this._cb[cmd.id];
+        if (result.error) throw new Error(result.error);
+        return result;
+      }
+
+    });
+
+    ["ls", "rm", "mkdir"].forEach(c => {
+
+      self[c] = async (path) => {
+        let cmd = Command[c](path);
+        let result = await self._writeAndResolve(cmd);
+        delete this._cb[cmd.id];
+        if (result.error) throw new Error(result.error);
+        return result;
+      }
+
+    });
+  }
+  
+  _writeAndResolve (cmd) {
+    let self = this;
+    return new Promise(resolve => {
+      self.stdin.write(cmd);
+      this._cb[cmd.id] = resolve;
+    });
+  }
+
+  * _putfile (dest) {  
```

Comment:
Also are we sure this should indeed return a promise?

I'm not sure, but couldn't it just return a writeable stream..and then have the stream emit `finish` when everything is flushed, and `error` if there is an error...
I could be wrong... I didn't think it through 100%.